contracts/minter/src/contract.rs

```diff
@@ -0,0 +1,602 @@
+use sg_std::{burn_and_distribute_fee, StargazeMsgWrapper, GENESIS_MINT_START_TIME, NATIVE_DENOM};
+use whitelist::msg::{
+    ConfigResponse as WhitelistConfigResponse, HasMemberResponse, QueryMsg as WhitelistQueryMsg,
+    UnitPriceResponse,
+};
+
+pub type Response = cosmwasm_std::Response<StargazeMsgWrapper>;
+pub type SubMsg = cosmwasm_std::SubMsg<StargazeMsgWrapper>;
+
+// version info for migration info
+const CONTRACT_NAME: &str = "crates.io:sg-minter";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+const INSTANTIATE_SG721_REPLY_ID: u64 = 1;
+
+// governance parameters
+const MAX_TOKEN_LIMIT: u32 = 10000;
+const MAX_PER_ADDRESS_LIMIT: u32 = 30;
+const STARTING_PER_ADDRESS_LIMIT: u32 = 5;
+const MIN_MINT_PRICE: u128 = 50_000_000;
+const MINT_FEE_PERCENT: u32 = 10;
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    if msg.num_tokens > MAX_TOKEN_LIMIT.into() {
+        return Err(ContractError::MaxTokenLimitExceeded {
+            max: MAX_TOKEN_LIMIT,
+        });
+    }
+
+    if let Some(per_address_limit) = msg.per_address_limit {
+        // Check per address limit is valid
+        if per_address_limit > MAX_PER_ADDRESS_LIMIT {
+            return Err(ContractError::InvalidPerAddressLimit {
+                max: MAX_PER_ADDRESS_LIMIT.to_string(),
+                got: per_address_limit.to_string(),
+            });
+        }
+    }
+
+    // Check that base_token_uri is a valid IPFS uri
+    let parsed_token_uri = Url::parse(&msg.base_token_uri)?;
+    if parsed_token_uri.scheme() != "ipfs" {
+        return Err(ContractError::InvalidBaseTokenURI {});
+    }
+
+    if NATIVE_DENOM != msg.unit_price.denom {
+        return Err(ContractError::InvalidDenom {
+            expected: NATIVE_DENOM.to_string(),
+            got: msg.unit_price.denom.to_string(),
+        });
+    }
+    if MIN_MINT_PRICE > msg.unit_price.amount.into() {
+        return Err(ContractError::InsufficientMintPrice {
+            expected: MIN_MINT_PRICE,
+            got: msg.unit_price.amount.into(),
+        });
+    }
+
+    // Initially set per_address_limit if no msg
+    let per_address_limit: Option<u32> = msg.per_address_limit.or(Some(STARTING_PER_ADDRESS_LIMIT));
+
+    let whitelist_addr: Option<Addr> = match msg.whitelist {
+        Some(wl) => Some(deps.api.addr_validate(&wl)?),
+        None => None,
+    };
+
+    // default is genesis mint start time
+    let default_start_time = Expiration::AtTime(Timestamp::from_nanos(GENESIS_MINT_START_TIME));
+    let start_time = match msg.start_time {
+        Some(st) => {
+            if st < default_start_time {
+                default_start_time
+            } else {
+                st
+            }
+        }
+        None => default_start_time,
+    };
```

Comment:
``` 
    let start_time = msg.start_time.map_or(default_start_time, |st|  max(st, default_start_time));
```

Reply:
I don't think `max` ordinal is implemented for `Expiration`. Will check (or contribute if not).